### PR TITLE
HLA-1150: MVM failure due to HAPProduct constructor mismatch

### LINKED ISSUE #1692: MVM failure due to missing log_level
_Issue [HLA-1150](https://jira.stsci.edu/browse/HLA-1150) was created on JIRA by [Steve Goldman](https://jira.stsci.edu/secure/ViewProfile.jspa?name=sgoldman):_

2023305083553 INFO src=drizzlepac.haputils.poller_utils- There are 5 usable images identified in the poller file for processing.
Traceback (most recent call last):
  File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3/envs/caldp_satandtools/lib/python3.9/site-packages/drizzlepac/hapmultisequencer.py", line 359, in run_mvm_processing
    obs_info_dict, total_obj_list = poller_utils.interpret_mvm_input(input_filename, log_level,
  File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3/envs/caldp_satandtools/lib/python3.9/site-packages/drizzlepac/haputils/poller_utils.py", line 332, in interpret_mvm_input
    obset_dict, tdp_list = parse_mvm_tree(obset_tree, all_mvm_exposures, log_level)
  File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3/envs/caldp_satandtools/lib/python3.9/site-packages/drizzlepac/haputils/poller_utils.py", line 512, in parse_mvm_tree
    sep_obj = SkyCellExposure(prop_id, obset_id, prod_list[1], prod_list[2],
TypeError: *_init_*() missing 1 required positional argument: 'log_level'

### PR DESCRIPTION
**Pull request by s-goldman**
Resolves [HLA-1150](https://jira.stsci.edu/browse/HLA-1150)

Closes #1692 

This PR addresses a failure in dlhstdmcv2 HST Auto-run for 10/31/23 where the following error was thrown:

`TypeError: init() missing 1 required positional argument: 'log_level'`

This appears linked [HLA-1110](https://jira.stsci.edu/browse/HLA-1110) (#1683) likely due to aperture not being added to the Product Objects for MVM processing. A test should be created to pick up these kinds of errors in the future. 

The [first commit](https://github.com/spacetelescope/drizzlepac/commit/ee3b64880ef93dbc0775d900f5214fde286345c5) shows the relevant changes for this PR, he second commit is the result of blacking poller_utils.py

Additional issues were found in the process of fixing these bugs. The code reduces the number of files to process using the analyze module and produces a new list called usable_datasets. This resulted in a uneven match of the new aperture column and the list that is passed through the code: cols. The new code now returns indices along with the usable_dataset, in order to select the correct aperture values to pass.

### CODE REVIEW

**Comment by s-goldman** [comment]
As of now, the code is creating MVMs with filenames e.g. 

hst_skycell-p1749x18y06_wfpc2_pc_empty_aperture_f502n_all_drz.fits

Looking for ways to remove the empty_aperture keyword

File: drizzlepac/haputils/poller_utils.py [hunk 1]
```diff
@@ -491,16 +491,19 @@ def parse_mvm_tree(det_tree, all_mvm_exposures, log_level):
                 prod_info = (filename[0] + " " + filetype).lower()
                 #
                 # mvm prod_info = 'skycell_p1234_x01y01 wfc3 uvis f200lp all 2009 1 drz'
+                
+                # prod_list is 0: proposal_id, 1:observation_id, 2:instrument, 3:detector,
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py]: It looks like the old comment here is not correct, and the new comment is using the SVM terminology, so it is also not correct.  The problem here is the MVM poller historically has not been parsed to collect quite the same information as the SVM poller, and the comment here was not updated long ago to contain the correct information. 

Line 495 (prod_list) should reflect the actual tokenized values from LIne 493 (prod_info).  Please fix the Line 495 comment as the values are skycell, instrument, detector, filters, exposure_layer, year_layer, old_or_new, and filetype (CTE corrected or not).

File: drizzlepac/haputils/poller_utils.py [hunk 2]
```diff
@@ -491,16 +491,19 @@ def parse_mvm_tree(det_tree, all_mvm_exposures, log_level):
...
 
                 if prod_list[5].strip() != '':
-                    layer = (prod_list[3], pscale, prod_list[4], prod_list[5])
+                    layer = (prod_list[3], pscale, prod_list[5], prod_list[6])
                 else:
-                    layer = (prod_list[3], pscale, prod_list[4])
+                    layer = (prod_list[3], pscale, prod_list[5])
 
                 ftype = prod_list[-1]
                 cellid = prod_list[0].split('-')[1]
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py]: Add a comment before prop_id that prop_id is really the projection cell number, and obset_id represents the X and Y skycell coordinates.  I understand now why this code confused you.  The terminology is SVM and the SkyCell* classes should have had a different constructor signature!

**Comment by s-goldman** [comment]
It looks like the MVM output drz filename is based on the SkyCellProduct.info string. While we had previously added aperture to this string, it isn't needed for MVMs and so it has been removed from only the info string for SkyCellProduct objects. 

This appears to have fixed the naming issue.

File: drizzlepac/haputils/poller_utils.py [hunk 3]
```diff
@@ -686,15 +798,35 @@ def parse_obset_tree(det_tree, log_level):
...
+                if (
+                    prod_list[6].lower().find("g") != -1
+                    or prod_list[6].lower().find("pr") != -1
+                ):
                     is_grism = True
                     filt_indx -= 1
-                    grism_sep_obj = GrismExposureProduct(prod_list[0], prod_list[1], prod_list[2],
-                                                         prod_list[3], prod_list[4], filename[1], prod_list[6],
-                                                         prod_list[7], log_level)
+                    grism_sep_obj = GrismExposureProduct(
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py:807]: The "aperture_from_poller" has been added to the class for the GrismExposureProduct, but the parameter is missing here.  It needs to be added here and NOT removed from the class.

File: drizzlepac/haputils/poller_utils.py [hunk 4]
```diff
@@ -1096,33 +1305,43 @@ def build_poller_table(input: str, log_level, all_mvm_exposures=[], poller_type=
...
-        if 'aperture' in input_table.colnames:
-            cols['aperture'] = input_table['aperture'].tolist()
+        if "aperture" in input_table.colnames:
+            cols["aperture"] = input_table["aperture"].tolist()
         else:
-            add_col = Column(['empty_aperture'] * len(usable_datasets), name='aperture', dtype='str')
+            add_col = Column(
+                ["empty_aperture"] * len(usable_datasets), name="aperture", dtype="str"
+            )
             input_table.add_column(add_col, index=7)
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py:1334]: There is a failure when trying to add a column when the number of usable datasets and the length of the input table are not the same.  This can be true for **_many_** reasons.   The job of the analyze module is to identify and return only the datasets which can actually be used during processing.

File: drizzlepac/haputils/poller_utils.py (hunk 3 above)
Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py:807]: Sorry. I might be a little confused. `prod_list[4]` is `aperture_from_poller` and should line up with the correct input.

File: drizzlepac/haputils/poller_utils.py (hunk 4 above)
Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py:1334]: add_col = Column(
                ["empty_aperture"] * len(input_table), name="aperture", dtype="str"
            )

File: drizzlepac/haputils/poller_utils.py (hunk 3 above)
Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py:807]: issue appears resolved.

File: drizzlepac/haputils/poller_utils.py (hunk 4 above)
Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py]: Issue should be resolved in [most recent commit](https://github.com/spacetelescope/drizzlepac/pull/1693/commits/511fa54a9808d8b5e8c8ef5a6ce0dfe395c5d8e9)

File: drizzlepac/haputils/poller_utils.py [hunk 5]
```diff
@@ -1017,44 +1197,69 @@ def build_poller_table(input: str, log_level, all_mvm_exposures=[], poller_type=
...
+                        f"Archived input image {table_line['pathname']} not found."
+                    )
                     err_msg = f"Input image {table_line['filename']} missing from current working directory and from the path specified in the poller file. Exiting... "
                     log.error(err_msg)
                     raise Exception(err_msg)
-        
-        elif (poller_type == 'mvm') & (len(input_table.columns) != len(poller_colnames)):
-            log.error(f"MVMs should use full poller files with {len(poller_colnames)} columns.")
+
+        elif (poller_type == "mvm") & (
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py:1246]: This is an old incorrect statement.  This should be a logical **and** versus a bitwise **&**!

File: drizzlepac/haputils/poller_utils.py [hunk 6]
```diff
@@ -1328,14 +1328,9 @@ def build_poller_table(
         if "aperture" in input_table.colnames:
             cols["aperture"] = input_table["aperture"].tolist()
         else:
-            add_col = Column(
-                ["empty_aperture"] * len(usable_datasets), name="aperture", dtype="str"
-            )
-            input_table.add_column(add_col, index=7)
-            poller_dtype += ("str",)
+            cols["aperture"] = ["empty_aperture"] * len(usable_datasets)
```
Comment (mdlpstsci) [inline drizzlepac/haputils/poller_utils.py:1331]: It is possible for cols["aperture"] Line 1331 to have more rows than are remaining in usable_datasets which will ultimately cause a failure.

Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py:1331]: It shoudn't? Are tests still coming back with uneven indices?

Comment (s-goldman) [reply drizzlepac/haputils/poller_utils.py:1333]: I was able to find the edge case that you were alluding to: Full poller table with datasets that are not usable. I altered the analyze.py module to return indices (as well as usable_datasets) to retrieve the corresponding aperture values from the input_table/poller_file.

**Comment by s-goldman** [comment]
Code appears to run as expected and is passing [jenkins regression tests](https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FDrizzlepac-Developers-Pull-Requests/detail/Drizzlepac-Developers-Pull-Requests/20/pipeline/324)

**Review by mdlpstsci — APPROVED** [review]
